File: tracker/discord.py
import logging
import time
from datetime import datetime

import requests

from .models import ProductReport

logger = logging.getLogger(__name__)

MAX_CONTENT_LENGTH = 2000
DISCORD_RATE_LIMIT_RETRY_DELAY = 5


def send_reports(webhook_url: str, reports: list[ProductReport], bot_name: str):
    """Send price reports to Discord, one message per product."""
    for report in reports:
        message = _format_message(report)
        _send_webhook(webhook_url, message, bot_name)
        time.sleep(1)  # Brief delay between messages


def _format_message(report: ProductReport) -> str:
    """Format a ProductReport into a Discord message following the skill's spec."""
    lines = []

    # Alert headers
    if report.is_deal:
        target = report.product.target_price
        lines.append(
            f"🚨 **DEAL FOUND at or below your ${target:.2f} target! BUY NOW!** 🚨"
        )
    if report.is_alert:
        lines.append(
            "🔥🔥🔥 **LOW PRICE ALERT — Lowest price in a while!** 🔥🔥🔥"
        )

    if lines:
        lines.append("")

    # Product name and timestamp
    lines.append(f"**{report.product.name}**")
    checked = report.checked_at.strftime("%B %d, %Y %I:%M %p")
    lines.append(f"Checked: {checked}")
    lines.append("")

    # TCGPlayer section
    if report.tcgplayer:
        tcp = report.tcgplayer
        lines.append("📊 **TCGPlayer**")
        price_parts = []
        if tcp.market_price is not None:
            price_parts.append(f"Market: ${tcp.market_price:.2f}")
        if tcp.low_price is not None:
            price_parts.append(f"Low: ${tcp.low_price:.2f}")
        if tcp.mid_price is not None:
            price_parts.append(f"Mid: ${tcp.mid_price:.2f}")
        if price_parts:
            lines.append(" | ".join(price_parts))
        if tcp.product_url:
            lines.append(f"[View on TCGPlayer]({tcp.product_url})")
        lines.append("")

    # Web search results section
    if report.web_prices:
        lines.append("🔍 **Lowest Found Online**")
        for wp in report.web_prices:
            alert_prefix = ""
            if _price_triggers_alert(wp.price, report):
                alert_prefix = "🔥 "
            lines.append(f"{alert_prefix}**${wp.price:.2f}** - [{wp.source}]({wp.url})")
        lines.append("")

    # Footer: target and MSRP
    footer_parts = []
    if report.product.target_price is not None:
        footer_parts.append(f"Target: ~${report.product.target_price:.2f}")
    if report.product.msrp is not None:
        footer_parts.append(f"MSRP: ${report.product.msrp:.2f}")
    if footer_parts:
        lines.append(" | ".join(footer_parts))

    # Errors
    if report.errors:
        lines.append("")
        lines.append("⚠️ " + "; ".join(report.errors))

    message = "\n".join(lines)

    # Truncate if too long for Discord
    if len(message) > MAX_CONTENT_LENGTH:
        message = message[: MAX_CONTENT_LENGTH - 20] + "\n... (truncated)"

    return message


def _price_triggers_alert(price: float, report: ProductReport) -> bool:
    """Check if a specific price should get the 🔥 emoji."""
    if report.product.msrp is not None:
        threshold = report.product.msrp * report.product.alert_threshold_pct
        if price < threshold:
            return True
    if report.product.target_price is not None and price <= report.product.target_price:
        return True
    return False


def _send_webhook(webhook_url: str, content: str, bot_name: str):
    """Send a message to a Discord webhook with retry on rate limit."""
    payload = {
        "username": bot_name,
        "content": content,
    }

    for attempt in range(3):
        try:
            resp = requests.post(webhook_url, json=payload, timeout=10)

            if resp.status_code == 204:
                logger.info("Discord message sent successfully")
                return

            if resp.status_code == 429:
                retry_after = resp.json().get("retry_after", DISCORD_RATE_LIMIT_RETRY_DELAY)
                logger.warning("Discord rate limited, retrying after %s seconds", retry_after)
                time.sleep(retry_after)
                continue

            resp.raise_for_status()

        except requests.RequestException as e:
            logger.error("Failed to send Discord message (attempt %d): %s", attempt + 1, e)
            if attempt < 2:
                time.sleep(2)

    logger.error("All Discord send attempts failed")

File: extension/lib/discord.js
export async function sendReports(webhookUrl, reports, botName) {
  for (const report of reports) {
    const message = formatMessage(report);
    await sendWebhook(webhookUrl, message, botName);
    // Brief delay between messages
    await new Promise((r) => setTimeout(r, 1000));
  }
}

export function formatMessage(report) {
  const lines = [];

  // Alert headers
  if (report.isDeal) {
    const target = report.product.targetPrice.toFixed(2);
    lines.push(`🚨 **DEAL FOUND at or below your $${target} target! BUY NOW!** 🚨`);
  }
  if (report.isAlert) {
    lines.push("🔥🔥🔥 **LOW PRICE ALERT — Lowest price in a while!** 🔥🔥🔥");
  }
  if (lines.length > 0) lines.push("");

  // Product name and timestamp
  lines.push(`**${report.product.name}**`);
  const now = new Date(report.checkedAt);
  const checked = now.toLocaleString("en-US", {
    month: "long",
    day: "numeric",
    year: "numeric",
    hour: "2-digit",
    minute: "2-digit",
  });
  lines.push(`Checked: ${checked}`);
  lines.push("");

  // TCGPlayer section
  if (report.tcgplayer) {
    const tcp = report.tcgplayer;
    lines.push("📊 **TCGPlayer**");
    const parts = [];
    if (tcp.marketPrice != null) parts.push(`Market: $${tcp.marketPrice.toFixed(2)}`);
    if (tcp.lowPrice != null) parts.push(`Low: $${tcp.lowPrice.toFixed(2)}`);
    if (tcp.midPrice != null) parts.push(`Mid: $${tcp.midPrice.toFixed(2)}`);
    if (parts.length > 0) lines.push(parts.join(" | "));
    if (tcp.productUrl) lines.push(`[View on TCGPlayer](${tcp.productUrl})`);
    lines.push("");
  }

  // Web search results
  if (report.webPrices && report.webPrices.length > 0) {
    lines.push("🔍 **Lowest Found Online**");
    for (const wp of report.webPrices) {
      const alertPrefix = priceTriggers(wp.price, report) ? "🔥 " : "";
      lines.push(`${alertPrefix}**$${wp.price.toFixed(2)}** - [${wp.source}](${wp.url})`);
    }
    lines.push("");
  }

  // Priority retailers — always shown, one line per retailer with status
  if (report.priorityResults && report.priorityResults.length > 0) {
    lines.push("📌 **Priority Retailers**");
    for (const pr of report.priorityResults) {
      if (pr.status === "available" && pr.price != null) {
        const alertPrefix = priceTriggers(pr.price, report) ? "🔥 " : "";
        const link = pr.url ? `[${pr.retailer}](${pr.url})` : pr.retailer;
        lines.push(`${alertPrefix}${link}: **$${pr.price.toFixed(2)}**`);
      } else if (pr.status === "sold_out") {
        lines.push(`${pr.retailer}: Sold out`);
      } else if (pr.status === "error") {
        lines.push(`${pr.retailer}: ⚠️ ${pr.message || "search error"}`);
      } else {
        lines.push(`${pr.retailer}: Not listed`);
      }
    }
    lines.push("");
  }

  // Footer
  const footer = [];
  if (report.product.targetPrice != null) {
    footer.push(`Target: ~$${report.product.targetPrice.toFixed(2)}`);
  }
  if (report.product.msrp != null) {
    footer.push(`MSRP: $${report.product.msrp.toFixed(2)}`);
  }
  if (footer.length > 0) lines.push(footer.join(" | "));

  // Errors
  if (report.errors && report.errors.length > 0) {
    lines.push("");
    lines.push("⚠️ " + report.errors.join("; "));
  }

  let message = lines.join("\n");
  if (message.length > 2000) {
    message = message.slice(0, 1980) + "\n... (truncated)";
  }
  return message;
}

function priceTriggers(price, report) {
  if (report.product.msrp != null) {
    const threshold = report.product.msrp * report.product.alertThresholdPct;
    if (price < threshold) return true;
  }
  if (report.product.targetPrice != null && price <= report.product.targetPrice) {
    return true;
  }
  return false;
}

async function sendWebhook(webhookUrl, content, botName) {
  const payload = {
    username: botName,
    content: content,
  };

  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const resp = await fetch(webhookUrl, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });

      if (resp.status === 204 || resp.status === 200) {
        console.log("Discord message sent successfully");
        return;
      }

      if (resp.status === 429) {
        const data = await resp.json();
        const retryAfter = data.retry_after || 5;
        console.warn(`Discord rate limited, retrying after ${retryAfter}s`);
        await new Promise((r) => setTimeout(r, retryAfter * 1000));
        continue;
      }

      console.error(`Discord error: ${resp.status}`);
    } catch (e) {
      console.error(`Discord send failed (attempt ${attempt + 1}):`, e);
      if (attempt < 2) await new Promise((r) => setTimeout(r, 2000));
    }
  }
  console.error("All Discord send attempts failed");
}
